###Installation on Mac OSX

You will need working knowledge of terminal and command line utilities in order to install/run funannotate.  By far the most challenging aspect is installing all of the dependencies correctly.  On Mac OSX, to install these tools we will rely extensively on HomeBrew.  Note HomeBrew is not actually a dependency of funannotate, but for this installation guide it is required.  You will also need to have `sudo` privileges to get all of these tools installed.

####Step-by-step instructions:

1) Install HomeBrew
```
ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install)‚Äù
```

2) Setup HomeBrew and Tap Repositories
```
#initial setup
brew doctor

#tap repositories
brew tap homebrew/science
brew tap nextgenusfs/tap
```

3) Install Python modules via pip (optionally install homebrew python)
```
#if system python requires sudo, you can perhaps get around by using homebrew local python
brew install python

#then setup pip and install modules to local python
pip install -U biopython natsort psutil goatools fisher numpy pandas matplotlib seaborn scikit-learn
```

4) Install Perl modules via cpanm (`brew install cpanm`)
```
cpanm Bio::Perl Getopt::Long Pod::Usage File::Basename threads threads::shared \
           Thread::Queue Carp Data::Dumper YAML Hash::Merge Logger::Simple Parallel::ForkManager \
           DBI Text::Soundex Scalar::Util::Numeric Tie::File POSIX Storable 
```
5) Install funannotation via homebrew
```
#update brew to make sure everything is current
brew update

#now install funannotate and dependencies
brew install funannotate

#you might need freetype library for matplotlib/seaborn
brew install freetype
```

6) Get RepBase data and reconfigure RepeatMasker/RepeatModeler. Register for [RepBase](http://www.girinst.org/repbase/)
```
#download RepeatMasker libraries and install
wget --user name --password pass http://www.girinst.org/server/RepBase/protected/repeatmaskerlibraries/repeatmaskerlibraries-20160829.tar.gz
tar zxvf repeatmaskerlibraries-20160829.tar.gz -C /usr/local/opt/repeatmasker/libexec

#now setup RepeatMasker
cd /usr/local/Cellar/repeatmasker/4.0.5/libexec
./configure

#softlink GFF script to bin in path
ln /usr/local/opt/repeatmasker/libexec/util/rmOutToGFF3.pl /usr/local/bin
```

7) Download and install GeneMark-ES/ET [here](http://exon.gatech.edu/GeneMark/license_download.cgi)
```
#uncompress and then move gmes_petap subdirectory to /usr/local
tar -xvf $HOME/Downloads/gm_et_macosx.tar
mv gm_et_macosx/gmes_petap/ /usr/local

#download key, then move to home directory
cp $HOME/Downloads/gm_key ~/.gm_key
```

8) Add several components and ENV variables to `~/.bash_profile` which will get sourced by your BASHRC - you can use any text editor for this
```
#add folders to PATH
export PATH="/usr/local/gmes_petap:$PATH"

#add environmental variables
export AUGUSTUS_CONFIG_PATH=/usr/local/opt/augustus/libexec/config
export EVM_HOME=/usr/local/opt/evidencemodeler
export GENEMARK_PATH=/usr/local/gmes_petap
export BAMTOOLS_PATH=/usr/local/opt/bamtools/bin
```

9) Re-launch a terminal window (or type `source ~/.bash_profile`). Finally run funannotate setup script to download databases and identify any problems.
```
#navigate into funannotate install directory
cd /usr/local/opt/funannotate/libexec

#run setup script, might need sudo here
./setup.sh
```

10) Troubleshooting.  There are a number of installation problems with a lot of these software packages that really bother me.  One common problem is that many of the programs written in perl ship with a shebang line of `#!/usr/bin/perl` - this can cause lots of problems if you are not using the system perl (which many people do to avoid messing with system perl as it is needed for lots of system maintenance).  I like to install perl using homebrew and install modules to this version of Perl, i.e. BioPerl, etc.  The better shebang line for portability is `#!/usr/bin/env perl` - which says to use whatever perl is currently in the environment, i.e. your homebrewed perl.  The same thing happens in python, i.e. the most portable is `#!/usr/bin/env python` - but that is not always the case.  There are several programs here that are by default installed to use system perl - if this is not what you have, you will have to do a little bit of extra work, here is the list of software I currently know has this problem.
1) GeneMark-ES
2) ProteinOrtho5
3) RepeatMasker
4) RepeatModeler

One solution is to manually change the shebang line, for example you can do this in the GeneMark folder as follows:
```
#move into folder
cd /usr/local/gmes_petap

#find all perl files, and change shebang line for each file inplace, using GNU-Sed here, Mac-sed may not work for this
find . -name "*.pl" | xargs gsed -i 's,#!/usr/bin/perl,#!/usr/bin/env perl,g'
```

The script will download and format necessary databases and then check all of the dependencies of funannotate - any tool not properly installed will be flagged by the script.

####Python Dependencies:
* Python 2
* Biopython
* psutil
* natsort
* goatools
* numpy
* pandas
* matplotlib
* seaborn
* sklearn library

####Software Dependencies:
* HomeBrew
* Perl
* BioPerl
* Blast+
* Hmmer3
* EVidenceModeler
* RepeatModeler
* RepeatMasker
* GMAP
* Blat - if using PASA results to train Augustus
* pslCDnaFilter (kent tools) - if using PASA results to train Augustus
* BedTools
* Augustus
* GeneMark-ES/ET (gmes_petap.pl)
* BamTools
* Genome Annotation Generator (gag.py)
* tbl2asn
* BRAKER1 (optional if training Augustus with RNA-seq data BAM file)
* Mummer
* RAxML
* mafft
* trimal

####Environmental variables:
EVM_HOME, GENEMARK_PATH, BAMTOOLS_PATH, AUGUSTUS_CONFIG_PATH
